Rationalised Signing Ceremony description (#52)
scionassociation/scion-cppki_I-D: draft-dekater-scion-pki.md

The participants of the Signing Ceremony fulfill different roles:
The Signing Ceremony SHOULD include the following participants:

- The **Ceremony Administrator** is in charge of moderating the signing process. The Ceremony Administrator guides all participants through the steps and may also act as an intermediary between participants when they share information with each other.
- A **Voting AS Representative** is capable of creating voting signatures on the TRC. The Voting Representative is in possession of a device with the private keys of the respective certificates in the TRC.
- A **Witness** is any person that participates in the ceremony as a passive entity. The Witness has no active role in any of the steps of the ceremony but can stop the process and ask for more information if they feel the integrity of the process might have been compromised.
- **Ceremony Administrator** - an individual in charge of moderating the signing process, guiding the participants through the steps, and acting as an intermediary for sharing information. The Ceremony Administrator is typically appointed by the ISD Manager or by resolution of the Voting ASes.

> **Note:** It is assumed that the member organizations of the ISD have decided in advance, before the Signing Ceremony, on the roles of the ceremony participants. That is, they have reached agreement about the Certificate Authority (CA) ASes (that will also issue the root certificates), the voting ASes, the voting AS representatives, the Ceremony Administrator and the Witnesses.

Author: I added this back.

- **Voting AS Representatives** - individuals representing each Voting AS who are able to create voting signatures on the TRC. They are in possession of a device with the private keys of their respective certificates in the TRC.

> **Note:** For the Signing Ceremony, it is assumed that all parties are trustworthy. Issues encountered during the ceremony are assumed to be caused by honest mistakes and not by malicious intent. Hash comparison checks are included to counter mistakes, such that every participant is sure that they operate on the same data. Furthermore, the private keys of each participant never leave their machine.The Ceremony Administrator does not have to be entrusted with private keys.

Author: Okay

The participants MUST decide in advance on the physical location of the Signing Ceremony, the devices that will be used, and the ISD policy as follows:

- ISD number - usually obtained from the SCION registry, see [](#id);
- The description of the TRC, see [](#description);
- Validity period of the TRC, see [](#validity);
- Voting quorum for the TRC, see [](#quorum);
- AS numbers of the Core ASes, see [](#core);
- AS numbers of the Authoritative ASes, see [](#auth);

> - The description and list of Control Plane root certificates.

Author: This was probably taken from the original text.

Each representative of a Voting AS MUST also create the following before the ceremony:

- A sensitive voting private key and a certificate holding the corresponding public key.
- A regular voting private key and a certificate holding the corresponding public key.

> In addition, each Certificate Issuer MUST create a control plane root private key and a certificate holding the corresponding public key. A representative of the Certificate Issuer need not be present at the ceremony as they do not need to sign the TRC, but they MUST provide their root certificate to be shared at the ceremony. The validity period of the certificates generated in advance MUST cover the full TRC validity period.

Reviewer: The old text also says 

> Representatives of CA ASes need not be present at the signing ceremony

But then it contraddicts itself in phase 1 of the ceremony: 

> The Ceremony Administrator now shares the bundle with the representatives of the voting and CA ASes.

We should resolve this contraddiction

Author: I'm not sure I see where the contradiction is.

Reviewer: If Representatives of CA ASes need not be present, then the Ceremony Administrator cannot share the bundle with them

Author: Does the bundle need to be shared with them at all, if they don't have to be present? Does the following work?

_The Ceremony Administrator now shares the bundle with the representatives of the voting ASes._

The Ceremony Administrator and Voting ASes MUST each bring to the Signing Ceremony a secure machine capable of signing and verifying TRCs, that has access to their own sensitive and regular voting private keys, and can compute the SHA-512 digest of the files.

The Ceremony Administrator MUST provide or be provided with a device to exchange data between the ceremony participants.

The Signing Ceremony should ideally include a procedure to verify that all devices are secure.

## Ceremony Process {ceremonyprocess}
{:numbered="false"}

The number of Voting ASes present at the Signing Ceremony must be equal to or larger than the specified voting quorum.

The signing process has four phases of data sharing, led by the Ceremony Administrator who provides instructions to the other participants:

### Phase 1: Certificate Exchange {#phase1}
{:numbered="false"}

All parties share the certificates that must be part of the TRC with the Ceremony Administrator. For the Voting ASes, these are the sensitive and the regular voting certificates, and for the Certificate Issuers these are the Control Plane root certificates.

> Each representative copies the requested certificates from their machine onto a data exchange device provided by the Ceremony Administrator that is passed between all representatives, before being returned to the Ceremony Administrator. Representatives must not copy their private keys onto the data exchange device as this invalidates the security of the ceremony.

Reviewer: ```suggestion
Each representative copies the requested certificates from their machine onto a data exchange device provided by the Ceremony Administrator that is passed between all representatives, before being returned to the Ceremony Administrator. Representatives MUST NOT copy the corresponding private keys onto the data exchange device as this invalidates the security of the ceremony.
```

The Ceremony Administrator then checks that the validity period of each provided certificate covers the previously agreed upon TRC validity, that the signature algorithms are correct, and that the certificate type is valid (root, sensitive voting or regular voting certificate). If these parameters are correct, the Ceremony Administrator computes the SHA256 hash value for each certificate, aggregates and bundles all the provided certificates, and finally calculates the SHA512 hash value for the entire bundle. All hash values must be displayed to the participants.

> The bundle MUST then be shared with all the representatives who must validate on their machine that the hash value of their certificates and that of the bundled certificates is the same as displayed by the Ceremony Administrator.

Reviewer: ```suggestion
The bundle MUST then be shared via a data exchange device with all the representatives of the voting and CA ASes, who must validate on their machine that the hash value of their certificates and that of the bundled certificates is the same as displayed by the Ceremony Administrator.
```

Phase 1 concludes when every representative has confirmed the SHA256 sums are correct. If there is any mismatch then this phase MUST be repeated.

### Phase 2: Generation of the TRC Payload {#phase2}
{:numbered="false"}

> The Ceremony Administrator generates the TRC payload based on the bundled certificates and the agreed ISD policy using a completed template with these values, see [](#ceremonyprep).

Reviewer: There is no template specified in #ceremonyprep . However, even in the previous version of the text of phase 2, there was also no clear definition of this template. I think we have to work more on this. Maybe we could add a reference to `3.1.2.2.   TRC Fields` ?

THis is actually referred to what is described in slide 9 of [this presentation](https://datatracker.ietf.org/meeting/118/materials/slides-118-panrg-scion-deployment-experience-the-secure-swiss-finance-network-ssfn-00) by SIX: 
<img width="909" alt="Screenshot 2025-02-19 at 12 20 10" src="https://github.com/user-attachments/assets/ff6e430c-1e58-4eec-bcfb-a9f35a831618" />

This is I think  a toml file like [this one](https://github.com/scionproto/scion/blob/ecb4eaa08a1ce7caf613b4da18f7ab2aecd04e46/scion-pki/conf/testdata/testcfg.toml#L8)

Author: I've changed this to _The Ceremony Administrator generates the TRC payload based on the bundled certificates and the TRC fields completed in accordance with ISD policy_ ...

> Once the voting representatives have verified the TRC data, the Ceremony Administrator computes the DER encoding of the data and the SHA256 hash value of the TRC payload file. The TRC payload file is then shared with the voting representatives via the data exchange device who verify the TRC payload hash value by computing this on their machine and checking it matches the one displayed by the Ceremony Administrator.

Reviewer: > the Ceremony Administrator computes the DER encoding of the data and 

Maybe we can add a reference to `TRC Schema `? 

```suggestion
Once the voting representatives have verified the TRC data, the Ceremony Administrator computes the DER encoding of the data according to [](#trcpayload) and the SHA256 hash value of the TRC payload file. The TRC payload file is then shared with the voting representatives via the data exchange device who verify the TRC payload hash value by computing this on their machine and checking it matches the one displayed by the Ceremony Administrator.
```

Phase 2 concludes when all voting representatives confirm that the contents of the TRC payload are correct.

### Phase 3: TRC Signing {#phase3}
{:numbered="false"}

> Each voting representative attaches a signature created with their private voting keys to the TRC (payload file), using their own machine. This serves to prove possession of the private keys.

Reviewer: ```suggestion
Each voting representative attaches a signature created with their own private voting key to the TRC (payload file), using their own machine. This serves to prove possession of the private keys.
```

Phase 3 concludes when all voting representatives have attached their signatures to the TRC.

### Phase 4: TRC Validation {#phase4}
{:numbered="false"}

> All voting representatives copy the TRC payload signed with their private voting keys to the data exchange device and return this to the Ceremony Administrator. The Ceremony Administrator assembles the final TRC by aggregating the payload data and verifying the signatures based on the certificates exchanged during Phase 1. The signed TRC is then shared with all the voting representatives who must again inspect the signatures.

Reviewer: I'd mention the data exchange device 

I'd also clarify the role of the administrator: 
```suggestion
All voting representatives copy the TRC payload signed with their private voting keys to the data exchange device and return this to the Ceremony Administrator. The Ceremony Administrator assembles the final TRC by aggregating the payload data and verifying the signatures based on the certificates exchanged during Phase 1. The Ceremony Administrator then shares the assembled TRC with all participants who MUST again inspect the signatures and verify them based on the certificates exchanged in Phase 1.
```
